crates/contracts-generate/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+        fmt::Write,
+        path::{Path, PathBuf},
+    },
+};
+
+const MOD_HEADER: &'static str = r#"#![allow(unused_imports, unused_attributes, clippy::all, rustdoc::all, non_snake_case)]
+    //! This module contains the sol! generated bindings for solidity contracts.
+    //! This is autogenerated code.
+    //! Do not manually edit these files.
+    //! These files may be overwritten by the codegen system at any time.
+    "#;
+
+pub struct Contracts {
+    pub contracts: Vec<Contract>,
+}
+
+impl Contracts {
+    pub fn new() -> Self {
+        Self { contracts: vec![] }
+    }
+
+    pub fn add_contract(mut self, contract: Contract) -> Self {
+        self.contracts.push(contract);
+        self
+    }
+
+    pub fn write_formatted(
+        self,
+        bindings_folder: &Path,
+        all_derives: bool,
+        output_folder: &Path,
+    ) -> eyre::Result<()> {
+        let mut mod_file = String::from(MOD_HEADER);
+
+        for contract in self.contracts {
+            let name = contract.name.clone();
+            contract.write_formatted(bindings_folder, all_derives, output_folder)?;
+            write_mod_name(&mut mod_file, &name)?;
+        }
+
+        let file: syn::File = syn::parse_file(&mod_file)?;
+        let formatted = prettyplease::unparse(&file);
+        std::fs::write(output_folder.join("mod.rs"), formatted)?;
+
+        Ok(())
+    }
+}
+
+pub struct Contract {
+    pub name: String,
+    networks: HashMap<u64, (String, Option<u64>)>,
+}
+
+impl Contract {
+    pub fn new<S: AsRef<str>>(name: S) -> Self {
+        Self {
+            name: name.as_ref().to_string(),
+            networks: HashMap::new(),
+        }
+    }
+
+    pub fn with_network<S: AsRef<str>>(
+        mut self,
+        id: u64,
+        (address, block_number): (S, Option<u64>),
+    ) -> Self {
+        self.networks
+            .insert(id, (address.as_ref().to_string(), block_number));
+        self
+    }
+
+    fn bindings_path(&self, bindings_folder: &Path) -> PathBuf {
+        bindings_folder.join(&self.name).with_extension("json")
+    }
+
+    pub fn generate<P: AsRef<Path>>(
+        self,
+        bindings_folder: P,
+        all_derives: bool,
+    ) -> eyre::Result<proc_macro2::TokenStream> {
+        let bindings_path = self.bindings_path(bindings_folder.as_ref());
+        let mut instance = SolMacroGen::new(bindings_path, self.name.clone());
+        generate_binding(&mut instance, all_derives)?;
+
+        let module_name_ident = format_ident!("{}", self.name);
+        let instance_name_ident = format_ident!("{}Instance", self.name);
+
+        let network_entries = self.networks.iter().map(|(id, (address, block_number))| {
+            let block_expr = match block_number {
+                Some(block) => quote::quote! { Some(#block) },
+                None => quote::quote! { None },
+            };
+            quote::quote! {
+                (#id, (::alloy::primitives::address!(#address), #block_expr))
```

Comment:
This currently converts the network names to u64 which makes the readability worse. Can we adjust this to use the named constants instead?